lib/provider/chromium.ts

```diff
@@ -0,0 +1,295 @@
+  osType = os.type();
+  osArch = os.arch();
+  outDir = OUT_DIR;
+  proxy: string = null;
+
+  constructor(providerConfig?: ProviderConfig) {
+    if (!providerConfig) {
+      return;
+    }
+    if (providerConfig.cacheFileName) {
+      this.cacheFileName = providerConfig.cacheFileName;
+    }
+    if (providerConfig.configFileName) {
+      this.configFileName = providerConfig.configFileName;
+    }
+    this.ignoreSSL = providerConfig.ignoreSSL;
+    if (providerConfig.osArch) {
+      this.osArch = providerConfig.osArch;
+    }
+    if (providerConfig.osType) {
+      this.osType = providerConfig.osType;
+    }
+    if (providerConfig.outDir) {
+      this.outDir = providerConfig.outDir;
+    }
+    if (providerConfig.proxy) {
+      this.proxy = providerConfig.proxy;
+    }
+  }
+
+  private makeDirectory(fileName: string) {
+    const dir = path.dirname(fileName);
+    try {
+      fs.mkdirSync(dir);
+    } catch (err) {
+    }
+  }
+
+  /**
+   * Step 1: Download the json file that contains all the releases by OS. Each
+   * OS will have a list of release versions. The requested body will also be
+   * written to the out directory.
+   * 
+   * The requested url is https://omahaproxy.appspot.com/all.json. Some other
+   * urls include a timestamped csv https://omahaproxy.appspot.com/history.
+   * @return Promise of the all-json file.
+   */
+  async downloadAllJson(): Promise<JsonObject> {
+    const fileName = path.resolve(this.outDir, this.cacheFileName);
+    if (!isExpired(fileName)) {
+      return JSON.parse(fs.readFileSync(fileName).toString());
+    } else {
+      this.makeDirectory(fileName);
+      const httpOptions = { fileName, ignoreSSL: this.ignoreSSL,
+        proxy: this.proxy };
+    
+      if (isExpired(fileName)) {
+        const allJsonUrl = 'https://omahaproxy.appspot.com/all.json';
+        let contents = await requestBody(allJsonUrl, httpOptions);
+        contents = `{ "all": ${contents} }`;
+        const jsonObj = JSON.parse(contents);
+        fs.writeFileSync(fileName, JSON.stringify(jsonObj, null, 2));
+        return jsonObj;
+      } else {
+        const contents = fs.readFileSync(fileName).toString();
+        return JSON.parse(contents);
+      }
+    }
+  }
+
+  /**
+   * Step 2: From the all-json object, make a request that matches the major
+   * version requested. The requested body will also be written to file in the
+   * out directory.
+   * 
+   * An example of a requsted url is
+   * https://omahaproxy.appspot.com/deps.json?version=72.0.3626.81
+   * @param allJson The all-json object.
+   * @param majorVersion The major version, this must be a whole number.
+   */
+  async downloadVersionJson(allJson: JsonObject, majorVersion: string
+      ): Promise<JsonObject> {
+    const fileName = path.resolve(this.outDir,
+      this.cacheVersionFileName.replace('.json', `-${majorVersion}.json`));
+    if (!isExpired(fileName)) {
+      return JSON.parse(fs.readFileSync(fileName).toString());
```

Comment:
Since this returns, you can drop the else clause and unnest everything below this point.

Reply:
Done.